### PR TITLE
Optimize nocapture encoding if only constant ptrs are stored

### PR DESCRIPTION
**Pull request by aqjune**
This is a simple patch that makes Memory::checkNocapture() return simply true if only constant pointers (null, global vars, undef) are stored into memory.

### CODE REVIEW

File: ir/globals.h [hunk 1]
```diff
@@ -72,7 +72,14 @@ extern bool has_null_block;
 /// Whether the programs do memory accesses that load/store int/ptrs
 extern bool does_int_mem_access;
 extern bool does_ptr_mem_access;
-extern bool does_ptr_store;
+
+struct PtrStoreInfo {
+  bool has_store;
+  bool constant_ptrs_only;
+  operator bool() const;
```
Comment (nunoplopes) [inline ir/globals.h:79]: you can inline this definition in the header

File: ir/globals.h [hunk 2]
```diff
@@ -72,7 +72,14 @@ extern bool has_null_block;
 /// Whether the programs do memory accesses that load/store int/ptrs
 extern bool does_int_mem_access;
 extern bool does_ptr_mem_access;
-extern bool does_ptr_store;
+
+struct PtrStoreInfo {
```
Comment (nunoplopes) [inline ir/globals.h:76]: why this complication (vs adding another bool)?
Is there other info you expect to collect from stored ptrs?

Comment (aqjune) [reply ir/globals.h:76]: Yes. I'm going to collect whether there is any local pointer that is stored, so a struct was used here.

Comment (aqjune) [reply ir/globals.h]: I'm not strongly in favor of having structs here, so fixed it to be const bools instead

**Comment by aqjune** [comment]
Ready to get review again

**Comment by nunoplopes** [comment]
is this any useful in practice? nocapture is not common at all.
Unless there's some interesting case, i would suggest we drop this patch and focus efforts on the new memory branch instead (after November).

**Comment by aqjune** [comment]
I think nocapture isn't rare - In case of sqlite3, among 1700 functions, about 500 functions have pointer arguments that are labeled as nocapture.
Storing a null pointer to somewhere seems to happen quite frequently, and attributes are attached at the early part of optimization pipeline, so I think this is helpful.

**Comment by aqjune** [comment]
LLVM unit test has improvement (1145 -> 1139 timeouts); maybe it is related to that nocapture was added to Alive2 because it appeared top from the list of unsupported features.

**Comment by nunoplopes** [comment]
I suggest we do pencils down now, and focus on the paper submission. We go back to memory afterwards.
My impression is that this patch (and a few other optimizations that we have currently) are not needed with the new memory branch. I can show you why later.

**Comment by aqjune** [comment]
Aha.. I see your point, capture analysis can be more precisely done in the new memory version.

**Comment by nunoplopes** [comment]
Right 😀

**Pull request closed without merging**